scripts/plugins/extauth-hook-AD.py

```diff
@@ -9,138 +9,268 @@
-
-from __future__ import print_function
+import abc
 import XenAPIPlugin
 import XenAPI
 import sys
 import os
-import getopt
 import tempfile
 import commands
-import syslog
+import logging
+import logging.handlers
+from enum import Enum
+
+
+def setup_logger():
+    logger = logging.getLogger()
+    logging.basicConfig(format='%(asctime)s %(levelname)s %(name)s %(funcName)s %(message)s', level=logging.DEBUG)
+    # Send to syslog local5, which will be redirected to xapi log /var/log/xensource.log
+    handler = logging.handlers.SysLogHandler(facility='local5', address='/dev/log')
+    std_err = logging.StreamHandler(sys.stderr)
+    logger.addHandler(handler)
+    logger.addHandler(std_err)
+
+
+setup_logger()
+logger = logging.getLogger(__name__)
+
+
+class ADBackend(Enum):
+    bd_pbis = 0
+    bd_winbind = 1
+
+
+class ADConfig(object):
+    def __init__(self, path, session, args, ad_enabled=True):
+        self._file_path = path
+        self._session = session
+        self._args = args
+        self._lines = []
+        self._backend = self._get_ad_backend(args)
+        self._ad_enabled = ad_enabled
+
+        if os.path.exists(self._file_path):
+            with open(self._file_path, 'r') as f:
+                self._lines = f.readlines()
+
+    def _get_ad_backend(self, args):
+        if  args.get("ad_backend", "winbind") == "pbis":
+            logger.debug("pbis is used as AD backend")
+            return ADBackend.bd_pbis
+
+        logger.debug("winbind is used as AD backend")
+        return ADBackend.bd_winbind
+
+    @abc.abstractmethod
+    def _apply_to_cache(self):
+        pass
 
-etc_pamd_sshd_start_boilerplate = """#%PAM-1.0
+    def apply(self):
+        self._apply_to_cache()
+        self._install()
+
+    def _install(self):
+        with tempfile.NamedTemporaryFile(prefix="extauth-", delete=False) as f:
+            lines = [line.encode() for line in self._lines]
+            f.writelines(lines)
+            os.rename(f.name, self._file_path)
+
+
+class StaticPam(ADConfig):
+    ad_pam_format = """#%PAM-1.0
 auth        required      pam_env.so
 auth        sufficient    pam_unix.so try_first_pass nullok
-auth        sufficient    /lib/security/pam_lsass.so try_first_pass
+auth        sufficient    {ad_module} try_first_pass try_authtok
 auth        required      pam_deny.so
 
 session     optional      pam_keyinit.so force revoke
 session     required      pam_limits.so
 session     [success=1 default=ignore] pam_succeed_if.so service in crond quiet use_uid
 session     required      pam_unix.so
 session     required      pam_loginuid.so
-session     sufficient    /lib/security/pam_lsass.so
+session     sufficient    {ad_module}
 
-account    required       pam_nologin.so
-account     required      /lib/security/pam_lsass.so unknown_ok
-# Start of list of allowed AD groups and users
-"""
+account     required      pam_nologin.so
+account     required      {ad_module} unknown_ok
+account     include       hcp_users
```

Comment:
I don't think we want to add this hcp_users here?

Reply:
> every subject (user or group) with Pool Admin role is added to /etc/pam.d/hcp_users as follows

so hcp_users is actually a list of pool admins, rather than a list of hcp users? calling the file `/etc/pam.d/pool_admins` would make more sense to me

Reply:
additionally do we want to add `account include hcp_users` in the pbis case?

Reply:
yes, we now we only add pool admin to this file, as we only permit pool admin to login ATM, 

yes, we do need to add `account include hcp_users` for pbis case, this is not specific to pbis or winbind, 
we should align both.